### PR TITLE
add docs about the route matching mechanism

### PR DESCRIPTION
**Pull request by aryszka**
document:

- routing mechanism with matching, request filters, backend request, response filters
- route lookup mechanism

### CODE REVIEW

File: docs/index.md [hunk 1]
```diff
@@ -58,3 +58,60 @@ another route, based on the changes that the filters made to the
...
 
 ![Skipper's request and response processing ](/img/req-and-resp-processing.svg)
+
+#### Routing mechanism
+
+The routing executes the following steps in the typical case:
+
+1. Select the best fitting route by matching the request against the
+   predicates. When no route found, respond with the default status code
+   (the default of the default is 404).
```
Comment (addityasingh) [inline docs/index.md]: probably a typo `(the default is 404)`

File: docs/index.md [hunk 2]
```diff
@@ -58,3 +58,60 @@ another route, based on the changes that the filters made to the
...
+4. Execute the filters defined in the route in reverse order on the
+   response. The filters may or may not alter the response.
+
+5. Respond to the incoming request with the resulting response.
+
+#### Route matching
+
+Skipper can handle a relatively large number of routes acceptable
+performance, while being able to use any attribute of the incoming HTTP
+requests to distinguis between them. In order to be able to do so, the
```
Comment (addityasingh) [inline docs/index.md]: typo `distinguish`

File: docs/index.md [hunk 3]
```diff
@@ -58,3 +58,60 @@ another route, based on the changes that the filters made to the
...
+   a response.
+
+4. Execute the filters defined in the route in reverse order on the
+   response. The filters may or may not alter the response.
+
+5. Respond to the incoming request with the resulting response.
+
+#### Route matching
+
+Skipper can handle a relatively large number of routes acceptable
```
Comment (addityasingh) [inline docs/index.md]: Missing **with**. `large number of routes with acceptable performance`

File: docs/index.md [hunk 4]
```diff
@@ -58,3 +58,60 @@ another route, based on the changes that the filters made to the
...
+performance, while being able to use any attribute of the incoming HTTP
+requests to distinguis between them. In order to be able to do so, the
+path matching predicates (Path() and PathSubtree() but not PathRegexp())
```
Comment (addityasingh) [inline docs/index.md]: nitpicking: Probably code formatting for the filter names would be better here and below

File: docs/index.md [hunk 5]
```diff
@@ -58,3 +58,60 @@ another route, based on the changes that the filters made to the
...
+   match the request, too, then the route with the most defined predicates
+   is matched.
+
+    * this is an O(n) lookup, but only on the same leaf
+
+    * the root of the tree is considered a single leaf, so if not using the
+      Path() or PathSubtree() predicates, the entire lookup will become O(n)
+      over all the routes.
+
+3. _If_ #2 results in multiple matching routes, then one route will be
```
Comment (addityasingh) [inline docs/index.md]: nitpicking: What about *then one route will be selected non-deterministically*?

File: docs/index.md (hunk 1 above)
Comment (aryszka) [reply docs/index.md]: kindof :) i'll try to phrase this in a different way

File: docs/index.md (hunk 5 above)
Comment (aryszka) [reply docs/index.md]: that would actually change the meaning. Stating unspecified leaves it more open for future changes without boundaries.

**Comment by addityasingh** [comment]
👍

**Comment by aryszka** [comment]
:+1:

**Pull request merged**